core/src/main/scala/org/apache/spark/rdd/PipedRDD.scala

```diff
@@ -112,73 +101,199 @@ private[spark] class PipedRDD[T: ClassTag](
-        if (lines.hasNext) {
-          true
-        } else {
-          val exitStatus = proc.waitFor()
-          if (exitStatus != 0) {
-            throw new Exception("Subprocess exited with status " + exitStatus)
+      private def cleanup(): Unit = {
+        // cleanup task working directory if used
+        if (workInTaskDirectory) {
+          scala.util.control.Exception.ignoring(classOf[IOException]) {
+            Utils.deleteRecursively(new File(taskDirectory))
           }
+          logDebug(s"Removed task working directory $taskDirectory")
+        }
+      }
 
-          // cleanup task working directory if used
-          if (workInTaskDirectory) {
-            scala.util.control.Exception.ignoring(classOf[IOException]) {
-              Utils.deleteRecursively(new File(taskDirectory))
-            }
-            logDebug("Removed task working directory " + taskDirectory)
-          }
+      private def propagateChildException(): Unit = {
+        val t = childThreadException.get()
+        if (t != null) {
+          val commandRan = command.mkString(" ")
+          logError("Caught exception while running pipe() operator. " +
+              s"Command ran: $commandRan.", t)
+          cleanup()
+          proc.destroy()
+          throw t
+        }
+      }
+    }
+  }
+}
+
+/** Specifies how to write the elements of the input [[RDD]] into the pipe. */
+trait InputWriter[T] extends Serializable {
+  def write(dos: DataOutput, elem: T): Unit
+}
 
-          false
+/** Specifies how to read the elements from the pipe into the output [[RDD]]. */
+trait OutputReader[T] extends Serializable {
+  /**
+   * Reads the next element. Could return null to indicate EOF.
+   *
+   * The input is guaranteed to have at least one byte.
+   */
+  def read(dis: DataInput): T
+}
+
+class TextInputWriter[I](
+    encoding: String = Codec.defaultCharsetCodec.name,
+    printPipeContext: (String => Unit) => Unit = null,
+    printRDDElement: (I, String => Unit) => Unit = null
+) extends InputWriter[I] {
+
+  private[this] val lineSeparator = System.lineSeparator().getBytes(encoding)
+  private[this] var initialized = printPipeContext == null
+
+  private def writeLine(dos: DataOutput, s: String): Unit = {
+    dos.write(s.getBytes(encoding))
+    dos.write(lineSeparator)
+  }
+
+  override def write(dos: DataOutput, elem: I): Unit = {
+    if (!initialized) {
+      printPipeContext(writeLine(dos, _))
+      initialized = true
+    }
+
+    if (printRDDElement == null) {
+      writeLine(dos, elem.toString)
+    } else {
+      printRDDElement(elem, writeLine(dos, _))
+    }
+  }
+}
+
+class TextOutputReader(
+    encoding: String = Codec.defaultCharsetCodec.name
+) extends OutputReader[String] {
+
+  private[this] val lf = "\n".getBytes(encoding)
+  private[this] val crlf = "\r\n".getBytes(encoding)
+  private[this] var buf = Array.ofDim[Byte](64)
+  private[this] var used = 0
+
+  @inline
+  private def endsWith(other: Array[Byte]): Boolean = {
```

Comment:
I did a quick benchmark of this version versus a generic one, here're the results.

```
Benchmark    Mode  Cnt    Score   Error   Units
handrolled  thrpt   20  170,990 ± 1,853  ops/us
stdlib      thrpt   20   37,940 ± 0,753  ops/us
```

Note that in the handrolled version we look at only the `used - 1` first elements of `buf`, therefore a fully equivalent generic version would be

```scala
java.util.Arrays.copyOfRange(buf, 0, used).endsWith(other)
```

which results in the following (here `bufSize` is the number of elements in `buf`)

```
Benchmark   (bufSize)   Mode  Cnt    Score   Error   Units
handrolled         64  thrpt   20  169,884 ± 3,538  ops/us
handrolled       1024  thrpt   20  171,257 ± 1,778  ops/us
generic            64  thrpt   20   27,226 ± 1,392  ops/us
generic          1024  thrpt   20   28,518 ± 0,362  ops/us
```

Combined with the fact that this function would be called for each character, I think it makes sense to keep it.